packages/talos_agent/src/agent.rs

```diff
@@ -0,0 +1,134 @@
+use async_trait::async_trait;
+use log::debug;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use tokio::sync::Notify;
+
+use crate::api::{AgentConfig, CertificationRequest, CertificationResponse, KafkaConfig, TalosAgent, TalosIntegrationType};
+use crate::messaging::api::{CandidateMessage, ConsumerType, Decision, DecisionMessage, PublisherType};
+use crate::messaging::kafka::KafkaConsumer;
+use crate::messaging::mock::MockConsumer;
+
+///
+/// Agent implementation which uses single consumer task. Once decision is received, it will lookup for
+/// pending in-flight transaction request by xid and notify it.
+///
+pub struct TalosAgentImpl {
+    pub config: AgentConfig,
+    pub kafka_config: Option<KafkaConfig>,
+    pub publisher: Box<PublisherType>,
+    // Key is xid, value is "shelf" where we keep notifier handle and exchanged value.
+    pub in_flight: Arc<Mutex<HashMap<String, Arc<InFlight>>>>,
+}
+
+pub struct InFlight {
+    pub xid: String,
+    // the exchanged value
+    pub decision: Mutex<Option<DecisionMessage>>,
+    pub monitor: Notify,
+}
+
+impl TalosAgentImpl {
+    /// Makes an instance of agent which is connected to kafka broker and already listening for incoming messages
+    pub fn new(config: AgentConfig, kafka_config: Option<KafkaConfig>, publisher: Box<PublisherType>) -> Self {
+        // todo:
+        //      The 'kafka_config' is needed to create a consumer from within task. Given that consumer is trait,
+        //      the outside builder should be creating Kafka or Mock consumer. Maybe it is better if task itself is
+        //      passed created by builder, as I am not yet sure how to move consumer into the task.
+        TalosAgentImpl {
+            config,
+            kafka_config,
+            publisher,
+            in_flight: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    fn create_kafka_consumer(agent_id: String, config: &KafkaConfig) -> Result<Box<ConsumerType>, String> {
+        let kc = KafkaConsumer::new(agent_id, config);
+        match kc.subscribe() {
+            Ok(()) => Ok(Box::new(kc)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn create_mock_consumer(_config: &KafkaConfig) -> Result<Box<ConsumerType>, String> {
+        Ok(Box::new(MockConsumer {}))
+    }
+
+    /// Starts listener task which reads Talos decisions from the topic
+    pub fn start(&self, int_type: &TalosIntegrationType) -> Result<(), String> {
+        let agent_id = self.config.agent_id.clone();
+        let config = self.kafka_config.clone().expect("Kafka configuration is required");
+        let in_flight = Arc::clone(&self.in_flight);
+
+        let it = int_type.clone();
+        tokio::spawn(async move {
+            let consumer: Box<ConsumerType> = match it {
```

Comment:
Would be nice in future iterations to be able to pass the Consumer based on `trait` and therefore I can pass the actual consumer or any mock to run my tests. 

Currently I can pass only the real consumer or a mock which will return `None` on receive message. 

Not needed in this PR, but would be great to have this refactored later.